### ISSUE TITLE
BUG: models with AutoEncoder layers not as first layer cause DisconnectedInputErrors on compilation

### ISSUE THREAD

**Issue by around1991** [fchollet/keras#1275]
Hi, I've got another bug report, sorry :(

The following sample code causes an error when compiling:

``` python
from keras.models import Sequential
encoder = Dense(input_dim=10, output_dim=2)
decoder = Dense(input_dim=2, output_dim=10)
model = Sequential()
model.add(Dense(input_dim=20, output_dim=10))
model.add(AutoEncoder(encoder=encoder, decoder=decoder, output_reconstruction=False))
model.compile(loss='mse', optimizer='sgd')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/auto/homes/kc391/event_rnn/env/bin/keras/keras/models.py", line 432, in compile
    train_loss)
  File "/auto/homes/kc391/event_rnn/env/bin/keras/keras/optimizers.py", line 79, in get_updates
    grads = self.get_gradients(loss, params)
  File "/auto/homes/kc391/event_rnn/env/bin/keras/keras/optimizers.py", line 47, in get_gradients
    grads = K.gradients(loss, params)
  File "/auto/homes/kc391/event_rnn/env/bin/keras/keras/backend/theano_backend.py", line 365, in gradients
    return T.grad(loss, variables)
  File "/auto/homes/kc391/event_rnn/env/bin/Theano/theano/gradient.py", line 545, in grad
    handle_disconnected(elem)
  File "/auto/homes/kc391/event_rnn/env/bin/Theano/theano/gradient.py", line 532, in handle_disconnected
    raise DisconnectedInputError(message)
theano.gradient.DisconnectedInputError: grad method was asked to compute the gradient with respect to a variable that is not part of the computational graph of the cost, or is used only by a non-differentiable operator: <TensorType(float32, matrix)>
Backtrace when the node is created:
  File "/auto/homes/kc391/event_rnn/env/bin/keras/keras/backend/theano_backend.py", line 34, in variable
    return theano.shared(value=value, name=name, strict=False)

```

The error is caused by:
- when AutoEncoder is inited, `self.params` is set by scraping the weights and biases from the encoder and decoder layers.
- when `model.add` is called with an AutoEncoder layer as the argument, this calls `AutoEncoder.set_previous()`, which calls `AutoEncoder.encoder.set_previous()`. 
- this reinitialises the weights and biases on the encoder layer, so that the pointers contained in `AutoEncoder.params` no longer points to the correct weights and biases on the encoder layer.

Possible solutions:
- update `AutoEncoder.params` when `AutoEncoder.set_previous()` is called
- add an ability to call `set_previous()` without overwriting existing weights (similar to #1266)

Thoughts?
Kris

**Comment by Sebubu** [comment]
+1 same here by adding 2 graphs in a model

**Comment by phreeza** [comment]
I believe this is probably also what broke my implementation of generative adversarial nets, recently reported by @iskandr in #1034

**Comment by Sebubu** [comment]
Simple example:

```python
from keras.models import Graph,  Sequential
from keras.layers.core import Dense
from keras.layers.normalization import BatchNormalization
from keras.layers.advanced_activations import PReLU


g = Graph()
g.add_input("input", input_shape=[20])
g.add_node(Dense(10), "dense", "input")
g.add_node(BatchNormalization(),"bn", "dense")
g.add_node(PReLU(),"activ", "bn")
g.add_output("output", "activ")
print "g output " + str(g.output_shape)

g2 = Graph()
g2.add_input("input", input_shape=[10])
g2.add_node(Dense(15), "dense", "input")
g2.add_node(BatchNormalization(),"bn", "dense")
g2.add_node(PReLU(),"activ", "bn")
g2.add_output("output", "activ")

model = Sequential()

model.add(g)
model.add(g2)

model.compile(loss="mse",optimizer="adadelta")
```

Error disappears as soon as I remove BatchNorm. Replacing PReLU or Dense makes no difference.

**Comment by Sebubu** [comment]
```python
from keras.models import Graph,  Sequential
from keras.layers.core import Dense
from keras.layers.normalization import BatchNormalization

g = Graph()
g.add_input("input1", input_shape=[20])
g.add_node(Dense(5), "dense1", "input1")
g.add_node(BatchNormalization(),"bn1", "dense1") 
g.add_output("output1", "bn1")

g2 = Graph()
g2.add_input("input2", input_shape=[5])
g2.add_node(Dense(15), "dense2", "input2")
g2.add_node(BatchNormalization(),"bn2", "dense2")
g2.add_output("output2", "bn2")

model = Sequential()
model.add(g)
model.add(g2)
```

BatchNorm (bn2) calls self.get_input() in his build method. Self.get_input() calls all layer below recursively and every layer stores the input of the previous layer in layer_cache. Thus Graph g2 has his input stored in his layer_cache. After adding g and g2 to the model, the cache is not refreshed and the model loss function is  calculated only for g2.
Can somebody confirm this?

I will do a pull request, although I don't see a point (without hacking) where the cache can be cleared right now.

**Comment by Sebubu** [comment]
Hint: Disabling layer_cache in each layer resolves one part of this issue. Unfortunately another exception appears now...

**Comment by berleon** [comment]
This issue is resolved by #1387 and should probably be closed
